allensdk/brain_observatory/behavior/behavior_ophys_session.py

```diff
@@ -1,98 +1,203 @@
+    `from_lims` and `from_nwb_path`.
     """
 
     @classmethod
-    def from_lims(cls, ophys_experiment_id):
+    def from_lims(cls, ophys_experiment_id: int) -> "BehaviorOphysSession":
         return cls(api=BehaviorOphysLimsApi(ophys_experiment_id))
 
     @classmethod
-    def from_nwb_path(cls, nwb_path, **api_kwargs):
-        api_kwargs["filter_invalid_rois"] = api_kwargs.get("filter_invalid_rois", True)
-        return cls(api=BehaviorOphysNwbApi.from_path(path=nwb_path, **api_kwargs))
+    def from_nwb_path(
+            cls, nwb_path: str, **api_kwargs: Any) -> "BehaviorOphysSession":
+        api_kwargs["filter_invalid_rois"] = api_kwargs.get(
+            "filter_invalid_rois", True)
+        return cls(api=BehaviorOphysNwbApi.from_path(
+            path=nwb_path, **api_kwargs))
 
     def __init__(self, api=None):
-
         self.api = api
 
-        self.ophys_experiment_id = LazyProperty(self.api.get_ophys_experiment_id)
-        self.max_projection = LazyProperty(self.get_max_projection)
-        self.stimulus_timestamps = LazyProperty(self.api.get_stimulus_timestamps)
-        self.ophys_timestamps = LazyProperty(self.api.get_ophys_timestamps)
-        self.metadata = LazyProperty(self.api.get_metadata)
-        self.dff_traces = LazyProperty(self.api.get_dff_traces)
-        self.cell_specimen_table = LazyProperty(self.api.get_cell_specimen_table)
-        self.running_speed = LazyProperty(self.api.get_running_speed)
-        self.running_data_df = LazyProperty(self.api.get_running_data_df)
-        self.stimulus_presentations = LazyProperty(self.api.get_stimulus_presentations)
-        self.stimulus_templates = LazyProperty(self.api.get_stimulus_templates)
-        self.licks = LazyProperty(self.api.get_licks)
-        self.rewards = LazyProperty(self.api.get_rewards)
-        self.task_parameters = LazyProperty(self.api.get_task_parameters)
-        self.trials = LazyProperty(self.api.get_trials)
-        self.corrected_fluorescence_traces = LazyProperty(self.api.get_corrected_fluorescence_traces)
-        self.average_projection = LazyProperty(self.get_average_projection)
-        self.motion_correction = LazyProperty(self.api.get_motion_correction)
-        self.segmentation_mask_image = LazyProperty(self.get_segmentation_mask_image)
+    # Using properties rather than initializing attributes to take advantage
+    # of API-level cache and not introduce a lot of overhead when the 
+    # class is initialized (sometimes these calls can take a while)
+    @property
+    def ophys_experiment_id(self) -> int:
+        """Unique identifier for this experimental session.
+        :rtype: int
+        """
+        return self.api.get_ophys_experiment_id()
+
+    @property
+    def max_projection(self) -> Image:
+        """2D max projection image.
+        :rtype: allensdk.brain_observatory.behavior.image_api.Image
+        """
+        return self.get_max_projection()
```

Comment:
Probably a dumb question, but how will this be cached as it does not appear to be a method of the api?

Would we instead implement the caching in `image_api`?

Reply:
I don't think we need to cache it. I just ran it and it was instantaneous.

Reply:
cool, sounds good to me then!

Reply:
Hm. Good call. No, let's add this to the API rather than handling it in the Session object. We want our cache to be consistent.

Reply:
I edited my original comment, so updating again. Looked into the methods and the `get_average_projection`, etc. calls on an API method under the hood to get the Image data (which are cached). The Session's instance method to just deserialize the image is pretty instantaneous and I don't think needs to be cached.